bn-apps/cordapp-updates-distribution/corda-updates-app/src/main/kotlin/net/corda/cordaupdates/app/member/SyncerService.kt

```diff
@@ -0,0 +1,98 @@
+package net.corda.cordaupdates.app.member
+
+import net.corda.businessnetworks.cordaupdates.core.ArtifactMetadata
+import net.corda.businessnetworks.cordaupdates.core.CordappSyncer
+import net.corda.businessnetworks.cordaupdates.core.SyncerConfiguration
+import net.corda.cordaupdates.transport.SessionConfigurationProperties
+import net.corda.core.node.AppServiceHub
+import net.corda.core.node.services.CordaService
+import net.corda.core.serialization.SingletonSerializeAsToken
+import java.io.File
+import java.util.concurrent.Callable
+import java.util.concurrent.Executors
+
+/**
+ * The cache that is populated by [SyncArtifactsFlow]. This cache is required to handle asynchronous invocations, when the results of the
+ * flow execution are not available immediately and are populated to the cache in the future instead.
+ *
+ * The contents of the cache are updated regardless of whether [SyncArtifactsFlow] was invoked asynchronously or not.
+ */
+@CordaService
+class ArtifactsMetadataCache(private val appServiceHub : AppServiceHub) : SingletonSerializeAsToken() {
```

Comment:
Does this need to be thread safe? If you're writing to `_cache` from one thread and reading from another, you do need explicit synchronisation, even if you think it's probably safe because the read happens after the right.

Reply:
The `_cache` is an immutable list and its contents never change. Once a new data arrives -I just swap the references. So probs we don't need synchronisation here.